StandardNodeGadget : Persistent nodule labels

ericmehl opened the pull request:
This adds the ability to add metadata to a node to either always show or always hide the node's nodule labels. If there's no metadata entry, the current behavior prevails and the labels are shown on hover.

I'm handling input and output plugs independently here - it seems like there could be cases where you'd want the labels on input plugs but not outputs (like a shader) or on output plugs but not inputs (like query nodes). It does make it a bit more verbose so happy to discuss if it would be better to settle on something simpler.

### Checklist ###

- [X] I have read the [contribution guidelines](https://github.com/GafferHQ/gaffer/blob/main/CONTRIBUTING.md).
- [X] I have updated the documentation, if applicable.
- [X] I have tested my change(s) in the test suite, and added new test cases where necessary.
- [X] My code follows the Gaffer project's prevailing coding style and conventions.

ericmehl commented:
Just noting I pushed an amendment that fixes the merge conflict, and makes a minor cosmetic change to a local variable name for consistency.

johnhaddon commented on src/GafferUI/StandardNodeGadget.cpp:
```diff
@@ -1133,9 +1137,24 @@ void StandardNodeGadget::enter( Gadget *gadget )
 		for( StandardNodule::RecursiveIterator it( gadget  ); !it.done(); ++it )
 		{
-			(*it)->setLabelVisible( true );
+			if(
+				(
+					(*it)->plug()->direction() == Plug::Direction::In &&
+					( !inputVisibleData || inputVisibleData->readable() )
+				) ||
+				(
+					(*it)->plug()->direction() == Plug::Direction::Out &&
+					( !outputVisibleData || outputVisibleData->readable() )
+				)
+			)
+			{
+				(*it)->setLabelVisible( true );
+			}
```
I'm not sure about this on two counts :

- I think it's more useful if `m_labelsVisibleOnHover` trumps `nodeGadget:inputNoduleLabelsVisible`. Without that you're almost completely in the dark as to what you're connecting to when `inputNoduleLabelsVisible` is False.
- `inputNoduleLabelsVisible == None` is acting like `inputNoduleLabelsVisible == True` here, but that's the opposite of how it acts in `updatePersistentNoduleLabels()`.

I think we get more useful and consistent behaviour if we just revert to the original code here.

ericmehl replied:
Reverted and squashed down into https://github.com/GafferHQ/gaffer/pull/6135/commits/7ec405e6f2e7789e74760eaf2695c9d70f2b811a (hopefully that's clearer than adding a new fixup to revert).

johnhaddon commented on src/GafferUI/StandardNodeGadget.cpp:
```diff
@@ -1158,10 +1192,20 @@ bool StandardNodeGadget::dragEnter( GadgetPtr gadget, const DragDropEvent &event
 		// Display the labels for all the compatible nodules so the
 		// user can see their options.
+		IECore::ConstBoolDataPtr inputVisibleData = Gaffer::Metadata::value<IECore::BoolData>( node(), g_inputNoduleLabelsVisibleKey );
+		IECore::ConstBoolDataPtr outputVisibleData = Gaffer::Metadata::value<IECore::BoolData>( node(), g_outputNoduleLabelsVisibleKey );
+
 		for( StandardNodule::RecursiveIterator it( this ); !it.done(); ++it )
 		{
-			(*it)->setLabelVisible( canConnect( event, it->get() ) );
+			if(
+				( (*it)->plug()->direction() == Plug::Direction::In && !inputVisibleData ) ||
+				( (*it)->plug()->direction() == Plug::Direction::Out && !outputVisibleData )
+			)
+			{
+				(*it)->setLabelVisible( canConnect( event, it->get() ) );
+			}
 		}
+
```
As for `enter()`, I think we'd be better off using the original code here. Currently we get no feedback at all when the metadata is specified (with either true or false). When the labels aren't visible by default, we need to see them now to have a chance of making a sensible connection. And when they are visible by default, it's useful to  hide the non-compatible ones so we get a clearer picture of the relevant options.

ericmehl replied:
Reverted and squashed into https://github.com/GafferHQ/gaffer/pull/6135/commits/7ec405e6f2e7789e74760eaf2695c9d70f2b811a

johnhaddon commented on src/GafferUI/StandardNodeGadget.cpp:
```diff
@@ -1222,14 +1263,32 @@ bool StandardNodeGadget::drop( GadgetPtr gadget, const DragDropEvent &event )
+{
+	if( auto standardNodule = IECore::runTimeCast<StandardNodule>( nodule ) )
+	{
+		bool inputVisible = false;
+		bool outputVisible = false;
+
+		if( IECore::ConstBoolDataPtr d = Gaffer::Metadata::value<IECore::BoolData>( node(), g_inputNoduleLabelsVisibleKey ) )
+		{
+			inputVisible = d->readable();
+		}
+		if( IECore::ConstBoolDataPtr d = Gaffer::Metadata::value<IECore::BoolData>( node(), g_outputNoduleLabelsVisibleKey ) )
+		{
+			outputVisible = d->readable();
+		}
```
We could test the plug direction first here, and then only look up the single bit of relevant metadata.

ericmehl replied:
Fixed in https://github.com/GafferHQ/gaffer/pull/6135/commits/31bf3651b2f9316f7fbc80631ed93f7f51c53415

johnhaddon commented on src/GafferUI/StandardNodule.cpp:
```diff
@@ -465,11 +467,21 @@ void StandardNodule::setCompatibleLabelsVisible( const DragDropEvent &event, boo
 	}
 
+	IECore::ConstBoolDataPtr inputVisibleData = Gaffer::Metadata::value<IECore::BoolData>( nodeGadget->node(), g_inputNoduleLabelsVisibleKey );
+	IECore::ConstBoolDataPtr outputVisibleData = Gaffer::Metadata::value<IECore::BoolData>( nodeGadget->node(), g_outputNoduleLabelsVisibleKey );
+
 	for( StandardNodule::RecursiveIterator it( nodeGadget ); !it.done(); ++it )
 	{
 		if( creator->canCreateConnection( it->get()->plug() ) )
 		{
-			(*it)->setLabelVisible( visible );
+			if( it->get()->plug()->direction() == Plug::Direction::In )
+			{
+				(*it)->setLabelVisible( !inputVisibleData ? visible : inputVisibleData->readable() );
+			}
+			if( it->get()->plug()->direction() == Plug::Direction::Out )
+			{
+				(*it)->setLabelVisible( !outputVisibleData ? visible : outputVisibleData->readable() );
+			}
```
As per my original comments, I think we can use the original code here too, for the same reasons as before.

ericmehl replied:
Without this, when you `dragLeave()` or `drop()` on a nodule, all the node's nodules visibility will be set to `false` because we're unconditionally doing that in those events.

Maybe instead of this, we create a private `staticVisibilityState()` (or somesuch name) that we use instead of `false` for the `dragLeave()` and `drop()` events?

johnhaddon replied:
Ah, OK. I had completely missed that `setCompatibleLabelsVisible()` was called for leave and drop as well as enter. Thanks for pointing that out.

> Maybe instead of this, we create a private staticVisibilityState() (or somesuch name) that we use instead of false for the dragLeave() and drop() events?

Yeah, or perhaps we just call `updatePersistentNoduleLabels()` on leave and drop? So the pattern is that helpful things set the visibility however they want, and then we always use `updatePersistentNoduleLabels()` to clobber everything back to the default state when we're done?

The problem with that of course is that `updatePersistentNoduleLabels()` is private to StandardNodeGadget at present. Which highlights how unsatisfactory the original code was in the first place, with the responsibility for visibility management spread out all over.

Maybe we just rename the private method to `StandardNodeGadget::applyNoduleLabelVisibilityMetadata()` and make it public?

ericmehl replied:
I went with the second solution in https://github.com/GafferHQ/gaffer/pull/6135/commits/a9f2a2fefeb86695074e5153a76ea1caa87fe6f4. And noting here our offline conversation where we decided to deprecate the current `updatePersistentNoduleLabels()` in favor of one that does not have the `visible` argument.

https://github.com/GafferHQ/gaffer/pull/6135/commits/a9f2a2fefeb86695074e5153a76ea1caa87fe6f4 maintains backward compatibility in case there are uses of `StandardNodule` we don't know about. Some of that can (must) be removed when we remove the deprecated `updatePersistentNoduleLabels()`.

johnhaddon reviewed (changes requested):
Thanks Eric, and sorry for once again taking far too long to getting to the review. I know how frustrating it is when code is no longer fresh in the mind when a review comes in.

Everything here seems well implemented, but I'd like to propose a slightly different set of principles for how the metadata combines with user interaction like hover/drag. Namely, the metadata provides the default "static" label visibility, and any hover/drag interaction completely overrides it to provide a "dynamic" label visibility that matches exactly what we had before. I think this makes for a better user experience, but also allows us to simplify the code, reducing the number of places that need to think about the metadata.

I've made a bunch of comments inline about what that might look like in practice - hopefully it makes sense...

Cheers...
John

ericmehl commented:
> the metadata provides the default "static" label visibility, and any hover/drag interaction completely overrides it to provide a "dynamic" label visibility that matches exactly what we had before

I believe then that setting the metadata entries to `False` won't change the behavior, since the default "static" value is `False`. Does that sound correct? Or are there cases where we do want to do something different with `False` set?

johnhaddon commented:
> I believe then that setting the metadata entries to `False` won't change the behavior, since the default "static" value is `False`. Does that sound correct? Or are there cases where we do want to do something different with `False` set?

To my mind, `False` should act exactly like `None` in all situations. When no metadata is registered (`None`), then conceptually we're just falling back to a default value of `False`. And that should behave identically to an explicitly registered value of `False`.

ericmehl commented on src/GafferUI/StandardNodeGadget.cpp:
```diff
@@ -1144,9 +1163,24 @@ void StandardNodeGadget::leave( Gadget *gadget )
+			)
+			{
+				(*it)->setLabelVisible( false );
+			}
```
Yes, we sure can - fixed in https://github.com/GafferHQ/gaffer/pull/6135/commits/77619080e91466037462d13419a95067a8fd3c61

ericmehl commented:
I addressed the comments above and I think it's working pretty nicely now. There's one detail that I'm not sure how to fix but perhaps can be ignored :

1. Put down a shader with the `ShaderPlugAdder` (+) "nodule" like a standard Arnold surface shader
2. Do `Gaffer.Metadata.registerValue( GafferArnold.ArnoldShader, "nodeGadget:inputNoduleLabelsVisible", True )` and `Gaffer.Metadata.registerValue( GafferArnold.ArnoldShader, "nodeGadget:outputNoduleLabelsVisible", True )` to set the in and out plug labels always visible
3. Put down an Arnold shader like `add`
4. Drag from the output of `add` over the standard surface nodules. All is well until you drag over the plug adder, when the "out" label on the standard surface shows up again. Or if you do the initial drag carefully enough to only hover over the "+" icon, the "out" label will still be visible.

This is because the `ShaderPlugAdder` is not handling `dragEnter` signals and related in order to show / hide the labels. Do we feel like that's something I should fix here?

ericmehl commented:
Talking offline, we decided to not worry about the `PlugAdder` hover, especially considering the somewhat awkward fix that would be needed.

My latest push made `applyNoduleLabelVisibilityMetadata()` private again, and made `StandardNodule` a friend of `StandardNodeGadget` so it can make use of that method.

I also added "Show Input Nodule Labels" and "Show Output Nodule Labels" to a new "Connections" sub-menu. The existing "Show Input / Output Connections" moved to that sub-menu too, in order to keep the menu from growing and growing.

Ready for a new look and hopefully a merge!

johnhaddon commented:
Thanks Eric! I've merged with one tweak to the naming of the menu items (Nodule is a weird term that I'm trying to keep out of the UI, and would like to replace elsewhere too).
